include/deal.II/base/scope_exit.h

```diff
@@ -0,0 +1,172 @@
+ *          }
+ *        catch (const RecoverableUserCallbackError &exc)
+ *          {
+ *            pending_exception = nullptr;
+ *            if (status == 0)
+ *              ; // just eat the exception
+ *            else
+ *              throw;
+ *          }
+ *        catch (...)
+ *          {
+ *            pending_exception = nullptr;
+ *            throw;
+ *          }
+ *      }
+ *    AssertKINSOL(status);
+ *
+ *    internal::copy(initial_guess_and_solution, solution);
+ *    // Free the vectors which are no longer used.
+ *#  ifdef DEAL_II_WITH_MPI
+ *    if (is_serial_vector<VectorType>::value == false)
+ *      {
+ *        N_VDestroy_Parallel(solution);
+ *        N_VDestroy_Parallel(u_scale);
+ *        N_VDestroy_Parallel(f_scale);
+ *      }
+ *    else
+ *#  endif
+ *      {
+ *        N_VDestroy_Serial(solution);
+ *        N_VDestroy_Serial(u_scale);
+ *        N_VDestroy_Serial(f_scale);
+ *      }
+ *    long nniters;
+ *    status = KINGetNumNonlinSolvIters(kinsol_mem, &nniters);
+ *    AssertKINSOL(status);
+ *    if (J != nullptr)
+ *      SUNMatDestroy(J);
+ *    if (LS != nullptr)
+ *      SUNLinSolFree(LS);
+ *    KINFree(&kinsol_mem);
+ *    return static_cast<unsigned int>(nniters);
+ * @endcode
+ *
+ * This code has a bug: At the bottom, it contains clean-up code
+ * that destroys a variety of objects created throughout the life of
+ * the function we are looking at. Because SUNDIALS is a library
+ * written in C, objects such as `solution`, `u_scale`, `J`, `LS`,
+ * etc., are really just regular pointers that need to be explicitly
+ * freed. But this does not happen in the places higher up where we
+ * `throw` an exception and don't catch it again immediately. One
+ * might be tempted to just copy the clean-up code to right before the
+ * `std::rethrow_exception` statement, but there is a path through the
+ * `catch` clauses marked with "just eat the exception" that ends up
+ * returning control flow back to the regular path, and if we hit this
+ * path, we would call the clean-up functions *twice* -- also not what
+ * we want. The upshot of this kind of example is that there are cases
+ * where it is quite difficult to say where exactly a function will
+ * exit and where the clean-up code needs to be placed; it is far
+ * easier to write it only once and let the compiler determine where
+ * it needs to be called, via a class such as the current one.
+ *
+ * The function is conceptually very similar to the
+ * [std::experimental::scope_exit](https://en.cppreference.com/w/cpp/experimental/scope_exit)
+ * class template that may find its way into a future C++ standard.
+ */
+class ScopeExit
+{
+public:
+  /**
+   * Constructor. Takes a function object that is to be executed at the
+   * place where the current object goes out of scope as argument.
+   */
+  ScopeExit(const std::function<void()> &exit_function);
+
+  /**
+   * Destructor. Execute the stored action.
+   */
+  ~ScopeExit();
+
```

Comment:
Would it make sense to `=delete` all four special member functions? Currently the class is copy- and move-constructible, which doesn't make a lot of sense and would likely not be what a user wants (and might even lead to a double delete).